src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs

```diff
@@ -226,6 +226,29 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(Func<string> cli
             return this;
         }
 
+        /// <summary>
+        /// Configures a delegate that creates a client assertion. See https://aka.ms/msal-net-client-assertion
+        /// </summary>
+        /// <param name="clientAssertionDelegate">delegate computing the client assertion used to prove the identity of the application to Azure AD.
+        /// This is a delegate that computes a Base-64 encoded JWT for each authentication call.</param>
+        /// <returns>The ConfidentialClientApplicationBuilder to chain more .With methods</returns>
+        /// <remarks> Callers can use this mechanism to cache their assertions </remarks>
+        public ConfidentialClientApplicationBuilder WithClientAssertion(Func<AssertionRequestOptions, string> clientAssertionDelegate)

```

Comment:
We should allow support for async, i.e. `Func<AssertionRequestOptions, Task<string>>`. Getting an assertion signed by KV is most likely an async operation.

Reply:
It supports both, see below. In the current solution there is both a sync and an async version. Just supported both in this proposed solution

Reply:
Not sure I understand this. If I want to implement a delegate like: 

```csharp
public string AsyncImplementation(AssertionRequestOptions options)
{
     string clientId = options.ClientId;
     string tokenEndpoint = options.TokenEndpoint;

     string json = CreateJson(clientId, tokenEndpoint);

     // async operation -> method must return a Task<string> not a string.
     string signedJson = await KeyVaultSignAsync(json);
     
    return signedJson;
}

```

Reply:
@bgavrilMS your call, do you want me to remove it or leave it to support the same as previously?

Reply:
"Async all the way" :). We've learned that the hard way. 

I can update the code as well if you wish.